lib/model/autocomplete.dart

```diff
@@ -188,8 +188,92 @@ class MentionAutocompleteView extends ChangeNotifier {
     required Narrow narrow,
   }) {
     assert(narrow is! CombinedFeedNarrow);
-    return store.users.values.toList()
-      ..sort((userA, userB) => compareByDms(userA, userB, store: store));
+    final users = store.users.values.toList();
+    switch (narrow) {
+      case StreamNarrow():
+        users.sort((userA, userB) => _compareByRelevance(
+          userA: userA,
+          userB: userB,
+          streamId: narrow.streamId,
+          topic: null,
+          store: store));
+      case TopicNarrow():
+        users.sort((userA, userB) => _compareByRelevance(
+          userA: userA,
+          userB: userB,
+          streamId: narrow.streamId,
+          topic: narrow.topic,
+          store: store));
+      case DmNarrow():
+        users.sort((userA, userB) => _compareByRelevance(
+          userA: userA,
+          userB: userB,
+          streamId: null,
+          topic: null,
+          store: store));
+      case CombinedFeedNarrow():
+        // do nothing in this case for now
+    }
+    return users;
+  }
+
+  static int _compareByRelevance({
+    required User userA,
+    required User userB,
+    required int? streamId,
+    required String? topic,
+    required PerAccountStore store,
+  }) {
+    if (streamId != null) {
+      final result = compareByRecency(
+        userA,
+        userB,
+        streamId: streamId,
+        topic: topic,
+        store: store);
+      if (result != 0) {
+        return result;
+      }
+    }
+    return compareByDms(userA, userB, store: store);
+  }
+
+  /// Determines which of the two users has more recent activity in the topic/stream.
+  ///
+  /// First checks for the activity in [topic] if provided.
+  ///
+  /// If no [topic] is provided, or the activity in the topic is the same (which
+  /// is extremely rare) or there is no activity in the topic at all, then
+  /// checks for the activity in the stream with [streamId].
+  ///
+  /// Returns a negative number if [userA] has more recent activity than [userB],
+  /// returns a positive number if [userB] has more recent activity than [userA],
+  /// and returns `0` if both [userA] and [userB] have the same recent activity
+  /// (which is extremely rare) or has no activity at all.
```

Comment:
Thank you for bringing this up. I tried to cover the two cases where comparing two message `id`s can result in `0`. One is that they're both `0` which means these two users sent no messages. Another is that they're both non-zero, but equal. At first, I thought that messages `id`s are based on the timestamp they're sent, so there's IS a very low probability they're sent at the exact same time. But now after looking at the Zulip API docs carefully, it seems that messages `id`s are unique. So maybe instead of "extremely rare", I can use the term "impossible". 😀